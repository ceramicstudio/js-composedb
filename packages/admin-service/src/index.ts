export { router } from './router.js'
export type { Router } from './router.js'
export { GraphQLQueryCodec, Service } from './service.js'
export type { GraphQLResult, ServiceConfig, ServiceParams } from './service.js'
