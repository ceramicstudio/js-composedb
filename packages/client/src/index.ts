/**
 * High-level ComposeDB client, based on the {@linkcode runtime ComposeDB runtime}.
 *
 * ## Installation
 *
 * ```sh
 * npm install @composedb/client
 * ```
 *
 * @module client
 */

export { ComposeClient, type ComposeClientParams } from './client.js'
