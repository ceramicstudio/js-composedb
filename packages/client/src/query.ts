import type {
  BaseQuery,
  CeramicApi,
  Page,
  Pagination,
  PaginationQuery,
  StreamState,
} from '@ceramicnetwork/common'
import { ModelInstanceDocument } from '@ceramicnetwork/stream-model-instance'
import type { Connection, ConnectionArguments } from 'graphql-relay'

export type ConnectionQuery = BaseQuery & ConnectionArguments

export function toBaseQuery(source: BaseQuery): BaseQuery {
  const query: BaseQuery = { model: source.model }
  if (source.account != null) {
    query.account = source.account
  }
  if (source.filter != null) {
    query.filter = source.filter
  }
  return query
}

export function toPaginationQuery(source: ConnectionQuery): PaginationQuery {
  const { after, before, first, last, ...base } = source

  let pagination: Pagination | undefined
  if (first != null) {
    pagination = { first }
    if (after != null) {
      pagination.after = after
    }
  } else if (last != null) {
    pagination = { last }
    if (before != null) {
      pagination.before = before
    }
  }
  if (pagination == null) {
    throw new Error('Missing "first" or "last" connection argument')
  }

  return { ...toBaseQuery(base), ...pagination }
}

export function toRelayConnection(
  ceramic: CeramicApi,
  page: Page<StreamState | null>
): Connection<ModelInstanceDocument | null> {
  return {
    edges: (page.edges ?? []).map(({ cursor, node }) => {
      return {
        cursor,
        node: node ? ceramic.buildStreamFromState<ModelInstanceDocument>(node) : null,
      }
    }),
    pageInfo: {
      ...page.pageInfo,
      startCursor: page.pageInfo.startCursor ?? null,
      endCursor: page.pageInfo.endCursor ?? null,
    },
  }
}

export async function queryConnection(
  ceramic: CeramicApi,
  query: ConnectionQuery
): Promise<Connection<ModelInstanceDocument | null>> {
  const indexQuery = toPaginationQuery(query)
  const page = await ceramic.index.query(indexQuery)
  return toRelayConnection(ceramic, page)
}

export async function querySingle(
  ceramic: CeramicApi,
  query: BaseQuery
): Promise<ModelInstanceDocument | null> {
  const indexQuery = { ...toBaseQuery(query), last: 1 }
  const result = await ceramic.index.query(indexQuery)
  const node = result.edges?.[0]?.node
  return node ? ceramic.buildStreamFromState<ModelInstanceDocument>(node) : null
}
