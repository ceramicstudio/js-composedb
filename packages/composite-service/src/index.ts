export { router } from './router.js'
export type { CreateModelInput, LoadModelInput, SaveCompositeInput, Router } from './router.js'
export { Service } from './service.js'
export type { GraphQLResult, SavedComposite, ServiceParams } from './service.js'
