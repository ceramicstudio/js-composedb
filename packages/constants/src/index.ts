export const NODE_ID_FIELD = 'id'

export const VIEWER_ID_HEADER = 'ComposeDB-ViewerID'
