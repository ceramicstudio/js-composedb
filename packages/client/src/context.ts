import type { BaseQuery, CeramicApi } from '@ceramicnetwork/common'
import type { ModelInstanceDocument } from '@ceramicnetwork/stream-model-instance'
import type { CommitID, StreamID } from '@ceramicnetwork/streamid'
import type { Connection } from 'graphql-relay'

import { type ConnectionQuery, queryConnection, querySingle } from './query.js'
import { type DocumentCache, DocumentLoader, type UpdateDocOptions } from './loader.js'

export type ContextParams = {
  /**
   * Optional cache for documents.
   */
  cache?: DocumentCache | boolean
  /**
   * Ceramic client instance.
   */
  ceramic: CeramicApi
  /**
   * @internal
   */
  loader?: DocumentLoader
}

/**
 * GraphQL execution context, exported by the {@linkcode client} module.
 *
 * ```sh
 * import { Context } from '@composedb/client'
 * ```
 */
export class Context {
  #ceramic: CeramicApi
  #loader: DocumentLoader

  constructor(params: ContextParams) {
    const { cache, ceramic } = params
    this.#ceramic = ceramic
    this.#loader = params.loader ?? new DocumentLoader({ ceramic, cache })
  }

  /**
   * Returns whether the Ceramic client instance used internally is authenticated or not. When not
   * authenticated, mutations will fail.
   */
  get authenticated(): boolean {
    return this.#ceramic.did?.authenticated ?? false
  }

  /**
   * Ceramic client instance used internally.
   */
  get ceramic(): CeramicApi {
    return this.#ceramic
  }

  /**
   * Document loader instance used internally.
   */
  get loader(): DocumentLoader {
    return this.#loader
  }

  /**
   * ID of the current viewer (authenticated DID), if set.
   */
  get viewerID(): string | null {
    const did = this.#ceramic.did
    return did?.hasParent ? did.parent : did?.id ?? null
  }

  /**
   * Load a document by ID, using the cache if possible.
   */
  async loadDoc<Content extends Record<string, any> = Record<string, any>>(
    id: string | CommitID | StreamID,
    fresh = false
  ): Promise<ModelInstanceDocument<Content>> {
    if (fresh) {
      this.#loader.clear(id)
    }
    return await this.#loader.load<Content>(id)
  }

  /**
   * Create a new document with the given model and content.
   */
  async createDoc<Content extends Record<string, any> = Record<string, any>>(
    model: string,
    content: Content
  ): Promise<ModelInstanceDocument<Content>> {
    return await this.#loader.create(model, content)
  }

  /**
   * Create a new single document with the given model and content.
   */
  async createSingle<Content extends Record<string, any> = Record<string, any>>(
    model: string,
    content: Content
  ): Promise<ModelInstanceDocument<Content>> {
    const controller = this.viewerID
    if (controller == null) {
      throw new Error('Document can only be created with an authenticated account')
    }
    const doc = await this.#loader.single<Content>(controller, model)
    await doc.replace(content)
    return doc
  }

  /**
   * Update an existing document.
   */
  async updateDoc<Content extends Record<string, any> = Record<string, any>>(
    id: string | StreamID,
    content: Content,
    options?: UpdateDocOptions
  ): Promise<ModelInstanceDocument<Content>> {
    return await this.#loader.update(id, content, options)
  }

  /**
   * Query the index for a connection of documents.
   */
  async queryConnection(query: ConnectionQuery): Promise<Connection<ModelInstanceDocument>> {
    return await queryConnection(this.#ceramic, query)
  }

  /**
   * Query the index for the total number of documents matching the query parameters.
   */
  async queryCount(query: BaseQuery): Promise<number> {
    return await this.#ceramic.index.queryCount(query)
  }

  /**
   * Query the index for a single document.
   */
  async querySingle(query: BaseQuery): Promise<ModelInstanceDocument | null> {
    return await querySingle(this.#ceramic, query)
  }
}
