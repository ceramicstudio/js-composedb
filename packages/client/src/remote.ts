import { fetchJson } from '@ceramicnetwork/common'
import { VIEWER_ID_HEADER } from '@composedb/constants'
import { createGraphQLSchema } from '@composedb/runtime'
import type { RuntimeCompositeDefinition } from '@composedb/types'
import { createBatchingExecutor } from '@graphql-tools/batch-execute'
import { stitchSchemas } from '@graphql-tools/stitch'
import type { Executor } from '@graphql-tools/utils'
import { type GraphQLSchema, print } from 'graphql'

export type GetViewerID = () => string | null | undefined

export function createRemoteExecutor(url: string, getViewerID: GetViewerID): Executor {
  const remoteExecutor: Executor = async ({ document, variables }) => {
    const viewerID = getViewerID()
    const headers = viewerID ? { [VIEWER_ID_HEADER]: viewerID } : {}
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return await fetchJson(url, {
      method: 'POST',
      headers,
      body: { query: print(document), variables },
    })
  }
  return createBatchingExecutor(remoteExecutor)
}

export type HybridSchemaParams = {
  definition: RuntimeCompositeDefinition
  getViewerID?: GetViewerID
  remoteExecutor?: Executor
  serverURL?: string
}

export function createHybridSchema(params: HybridSchemaParams): GraphQLSchema {
  const { definition, getViewerID, remoteExecutor, serverURL } = params

  let executor: Executor
  if (typeof remoteExecutor === 'function') {
    executor = remoteExecutor
  } else if (typeof getViewerID === 'function' && typeof serverURL === 'string') {
    executor = createRemoteExecutor(serverURL, getViewerID)
  } else {
    throw new Error(
      'Invalid parameters to create hybrid schema: missing `remoteExecutor` or `getViewerID` and `serverURL` parameters'
    )
  }
  const remoteSchemaConfig = {
    executor,
    schema: createGraphQLSchema({ definition, readonly: true }),
  }

  return stitchSchemas({
    subschemas: [createGraphQLSchema({ definition }), remoteSchemaConfig],
    mergeTypes: false,
    onTypeConflict: (left, right, info) => {
      return info?.left.subschema === remoteSchemaConfig ? left : right
    },
  })
}
