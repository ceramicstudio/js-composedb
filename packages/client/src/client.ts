import type { CeramicApi } from '@ceramicnetwork/common'
import { CeramicClient } from '@ceramicnetwork/http-client'
import {
  ComposeRuntime,
  type ComposeRuntimeParams,
  type Context,
  type DocumentCache,
  createContext,
} from '@composedb/runtime'
import type { RuntimeCompositeDefinition } from '@composedb/types'
import type { Executor } from '@graphql-tools/utils'
import type { DID } from 'dids'
import type { DocumentNode, ExecutionResult, Source } from 'graphql'

import { createHybridSchema } from './remote.js'

export type ComposeClientParams = {
  /**
   * Optional cache for documents.
   */
  cache?: DocumentCache
  /**
   * Ceramic client instance or HTTP URL.
   */
  ceramic: CeramicApi | string
  /**
   * Runtime composite definition, created using the {@linkcode devtools.Composite Composite}
   * development tools.
   */
  definition: RuntimeCompositeDefinition
  /**
   * Optional remote query executor.
   */
  remoteExecutor?: Executor
  /**
   * Optional {@linkcode server query server} URL.
   */
  serverURL?: string
}

/**
 * The ComposeClient class provides APIs to execute queries on a GraphQL schema generated from a
 * `RuntimeCompositeDefinition`, leveraging the {@linkcode runtime.ComposeRuntime ComposeRuntime class}.
 *
 * It is exported by the {@linkcode client} module.
 *
 * ```sh
 * import { ComposeClient } from '@composedb/client'
 * ```
 */
export class ComposeClient {
  #context: Context
  #resources: Array<string>
  #runtime: ComposeRuntime

  constructor(params: ComposeClientParams) {
    const { ceramic, definition, remoteExecutor, serverURL, ...contextParams } = params
    const ceramicClient = typeof ceramic === 'string' ? new CeramicClient(ceramic) : ceramic
    this.#context = createContext({ ...contextParams, ceramic: ceramicClient })
    this.#resources = Object.values(definition.models).map((model) => {
      return `ceramic://*?model=${model.id}`
    })
    const runtimeParams: ComposeRuntimeParams = {
      ceramic: ceramicClient,
      context: this.#context,
      definition,
    }
    if (remoteExecutor != null || serverURL != null) {
      runtimeParams.schema = createHybridSchema({
        definition,
        getViewerID: () => this.id,
        remoteExecutor,
        serverURL,
      })
    }
    this.#runtime = new ComposeRuntime(runtimeParams)
  }

  /**
   * Context instance used internally.
   */
  get context(): Context {
    return this.#context
  }

  /**
   * DID instance used internally by the Ceramic client instance.
   */
  get did(): DID | undefined {
    return this.#context.ceramic.did
  }

  /**
   * ID of the DID attached to the Ceramic client instance used internally. If `null`, the
   * Ceramic instance is not authenticated and mutations will fail.
   */
  get id(): string | null {
    return this.#context.getViewerID()
  }

  /**
   * CACAO resources URLs for the models the client interacts with.
   */
  get resources(): Array<string> {
    return this.#resources
  }

  /**
   * Attach the given DID instance to the Ceramic client instance used internally. An authenticated
   * DID instance is necessary to perform GraphQL mutations.
   */
  setDID(did: DID): void {
    this.#context.ceramic.did = did
  }

  /**
   * Execute a GraphQL query from a DocumentNode and optional variables.
   */
  async execute<Data = Record<string, unknown>>(
    document: DocumentNode,
    variableValues?: Record<string, unknown>
  ): Promise<ExecutionResult<Data>> {
    return await this.#runtime.execute<Data>(document, variableValues)
  }

  /**
   * Execute a GraphQL query from its source and optional variables.
   */
  async executeQuery<Data = Record<string, unknown>>(
    source: string | Source,
    variableValues?: Record<string, unknown>
  ): Promise<ExecutionResult<Data>> {
    return await this.#runtime.executeQuery<Data>(source, variableValues)
  }
}
