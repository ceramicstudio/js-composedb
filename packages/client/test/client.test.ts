/**
 * @jest-environment composedb
 */

import { Composite } from '@composedb/devtools'
import {
  createCommentSchemaWithPost,
  loadPostSchemaWithComments,
  postSchema,
  profilesSchema,
} from '@composedb/test-schemas'
import type { Executor } from '@graphql-tools/utils'
import { jest } from '@jest/globals'

import { ComposeClient } from '../src'

describe('client', () => {
  test('create profile', async () => {
    const composite = await Composite.create({ ceramic, schema: profilesSchema })
    const client = new ComposeClient({ ceramic, definition: composite.toRuntime() })
    const res = await client.executeQuery<{ createGenericProfile: { document: { id: string } } }>(
      `
      mutation CreateProfile($input: CreateGenericProfileInput!) {
        createGenericProfile(input: $input) {
          document {
            id
          }
        }
      }
      `,
      { input: { content: { name: 'Alice' } } }
    )
    expect(res.data?.createGenericProfile.document.id).toBeDefined()
  }, 30000)

  test('create and query post with comments', async () => {
    const postComposite = await Composite.create({ ceramic, schema: postSchema })
    const postModelID = postComposite.modelIDs[0]

    const commentComposite = await Composite.create({
      ceramic,
      schema: createCommentSchemaWithPost(postModelID),
    })
    const commentModelID = commentComposite.modelIDs.find((id) => id !== postModelID) as string

    const composite = await Composite.create({
      ceramic,
      schema: loadPostSchemaWithComments(postModelID, commentModelID),
    })
    const definition = composite.toRuntime()

    const client = new ComposeClient({ ceramic, definition })

    const createPostMutation = `
      mutation CreatePost($input: CreatePostInput!) {
        createPost(input: $input) {
          document {
            id
          }
        }
      }
    `
    await client.executeQuery(createPostMutation, {
      input: { content: { title: 'A first post', text: 'First post content' } },
    })
    const postRes = await client.executeQuery<{ createPost: { document: { id: string } } }>(
      createPostMutation,
      { input: { content: { title: 'A second post', text: 'Second post content' } } }
    )
    const postID = postRes.data?.createPost.document.id

    const createCommentMutation = `
      mutation CreateComment($input: CreateCommentInput!) {
        createComment(input: $input) {
          document {
            id
          }
        }
      }
    `
    await client.executeQuery(createCommentMutation, {
      input: { content: { postID, text: 'A first comment' } },
    })
    await client.executeQuery(createCommentMutation, {
      input: { content: { postID, text: 'A second comment' } },
    })
    await client.executeQuery(createCommentMutation, {
      input: { content: { postID, text: 'A third comment' } },
    })

    const res = await client.executeQuery(
      `
      query {
        viewer {
          postList(first: 5) {
            edges {
              node {
                title
                text
                commentsCount
                comments(first: 5) {
                  edges {
                    node {
                      text
                    }
                  }
                }
              }
            }
          }
        }
      }
      `
    )
    expect(res).toMatchSnapshot()
  }, 60000)

  test('with remote executor', async () => {
    const remoteExecutor = jest.fn(() => {
      return { data: {} }
    }) as Executor
    const composite = await Composite.create({ ceramic, schema: profilesSchema })
    const client = new ComposeClient({ ceramic, definition: composite.toRuntime(), remoteExecutor })
    const res = await client.executeQuery(`
      query {
        viewer {
          genericProfile {
            name
          }
        }
      }
    `)
    expect(res.errors).not.toBeDefined()
    expect(remoteExecutor).toHaveBeenCalled()
  }, 30000)
})
