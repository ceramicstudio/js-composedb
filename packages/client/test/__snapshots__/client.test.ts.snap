// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client create and query post with comments 1`] = `
{
  "data": {
    "viewer": {
      "postList": {
        "edges": [
          {
            "node": {
              "comments": {
                "edges": [],
              },
              "commentsCount": 0,
              "text": "First post content",
              "title": "A first post",
            },
          },
          {
            "node": {
              "comments": {
                "edges": [
                  {
                    "node": {
                      "text": "A first comment",
                    },
                  },
                  {
                    "node": {
                      "text": "A second comment",
                    },
                  },
                  {
                    "node": {
                      "text": "A third comment",
                    },
                  },
                ],
              },
              "commentsCount": 3,
              "text": "Second post content",
              "title": "A second post",
            },
          },
        ],
      },
    },
  },
}
`;
